fendermint/app/src/ipc.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! IPC related execution
+
+use crate::app::{AppState, AppStoreKey};
+use crate::{App, BlockHeight};
+use bytes::Bytes;
+use fendermint_rpc::response::decode_fevm_invoke_bytes;
+use fendermint_storage::{Codec, Encode, KVReadable, KVStore, KVWritable};
+use fendermint_vm_actor_interface::{ipc, system};
+use fendermint_vm_interpreter::fvm::state::FvmStateParams;
+use fendermint_vm_interpreter::fvm::FvmMessage;
+use fendermint_vm_topdown::convert::{
+    decode_parent_finality_return, encode_get_latest_parent_finality,
+};
+use fendermint_vm_topdown::sync::ParentFinalityStateQuery;
+use fendermint_vm_topdown::{IPCParentFinality, ParentFinalityProvider};
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::RawBytes;
+use fvm_shared::econ::TokenAmount;
+use num_traits::Zero;
+
+/// Queries the LATEST COMMITTED parent finality from the storage
+pub struct AppParentFinalityQuery<DB, SS, S, I, P>
+where
+    SS: Blockstore + 'static,
+    S: KVStore,
+{
+    /// The app to get state
+    app: App<DB, SS, S, I, P>,
+}
+
+impl<DB, SS, S, I, P> AppParentFinalityQuery<DB, SS, S, I, P>
+where
+    S: KVStore
+        + Codec<AppState>
+        + Encode<AppStoreKey>
+        + Encode<BlockHeight>
+        + Codec<FvmStateParams>,
+    DB: KVWritable<S> + KVReadable<S> + 'static + Clone,
+    SS: Blockstore + 'static + Clone,
+    P: ParentFinalityProvider + Send + Sync,
+{
+    pub fn new(app: App<DB, SS, S, I, P>) -> Self {
+        Self { app }
+    }
+}
+
+impl<DB, SS, S, I, P> ParentFinalityStateQuery for AppParentFinalityQuery<DB, SS, S, I, P>
+where
+    S: KVStore
+        + Codec<AppState>
+        + Encode<AppStoreKey>
+        + Encode<BlockHeight>
+        + Codec<FvmStateParams>,
+    DB: KVWritable<S> + KVReadable<S> + 'static + Clone,
+    SS: Blockstore + 'static + Clone,
+    P: ParentFinalityProvider + Send + Sync,
+{
+    fn get_latest_committed_finality(&self) -> anyhow::Result<Option<IPCParentFinality>> {
+        let maybe_exec_state = self.app.new_read_only_exec_state()?;
```

Comment:
You have to wonder if we could instantiate this `AppParentFinalityQuery` with a closure, something like 

```rust
pub struct AppParentFinalityQuery<F, DB>
where
    F: Fn() -> anyhow::Result<Option<FvmExecState<DB>>>
{
    get_exec_state: F
}

AppParentFinalityQuery::new(|| app.new_read_only_exec_state())
```

Way less code needs to be duplicated in the generic constraints.

Reply:
nice tip, but this might be called repeatedly, so still have to hold the `App`.

Reply:
You would still be able to hold the app if you used 

```rust
let qapp = app.clone()
AppParentFinalityQuery::new(move || qapp.new_read_only_exec_state())
```

The closure takes ownership of the app. But the signature is much simpler as you don't need to tell.